bpf: add metrics for fragmented ipv4 packets

jibi opened the pull request:
This commit introduces 2 new metrics in the datapath logic related to
fragmented IPv4 packets:

* `REASON_FRAG_PACKET`: number of received fragmented packets
* `REASON_FRAG_PACKET_UPDATE`: number of failures in updating the
`IPV4_FRAG_DATAGRAMS_MAP` map to register the first logical fragment of
a datagram

Regarding testing: I did a smoke test by running `make build_all` and I'm now waiting for e2e tests to finish

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -86,14 +87,18 @@ static __always_inline bool ipv4_is_not_first_fragment(const struct iphdr *ip4)
 
 static __always_inline int
-ipv4_frag_get_l4ports(const struct ipv4_frag_id *frag_id,
+ipv4_frag_get_l4ports(struct __ctx_buff *ctx,
```
```suggestion
ipv4_frag_get_l4ports(const struct __ctx_buff *ctx,
```

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -111,7 +116,11 @@ ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
 
-	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
+	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY)
+	    < 0)
```
```suggestion
	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY) < 0)
```
More readable on one line? Or break inside `map_update_elem()` otherwise?

jibi replied:
I assumed `checkpatch` would complain on lines with > 80 chars but the limit is actually set to 100 so I think  we are good to go with the one line option :+1:

qmonnet replied:
Right, it's still recommended to keep lines “short”, but the limit for checkpatch complaints was raised in the kernel repo a few months ago. In that case I believe it does read better on one line.

qmonnet replied:
Also for your information, we have a few false positives with checkpatch at the moment, that I intend to address. There's a list on https://github.com/cilium/cilium/issues/12467.

qmonnet replied:
Re-reading that line -- You could as well remove the `< 0`, since the helper always returns `0` on success.

qmonnet reviewed (changes requested):
Looks good to me, with a few nitpicks and pending the unit test build is fixed :)

qmonnet reviewed (changes requested):
Oh and you probably want to add a description in `pkg/monitor/api/drop.go` for the new metrics.

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -86,14 +87,18 @@ static __always_inline bool ipv4_is_not_first_fragment(const struct iphdr *ip4)
+	if (!tmp) {
+		update_metrics(ctx_full_len(ctx), METRIC_INGRESS,
+			       -DROP_FRAG_NOT_FOUND);
 		return DROP_FRAG_NOT_FOUND;
```
When we return an error, we will usually end up dropping the packet and calling `send_drop_notify_error`, which itself calls `send_drop_notify`:
https://github.com/cilium/cilium/blob/1843b4adeddad7ee0fd83bc3c71ba6e4d57db48e/bpf/lib/drop.h#L72-L83
So in this case, we will register two events for `DROP_FRAG_NOT_FOUND`, for a single packet.

I tried following the return path and it looks like in some cases, we simply ignore `DROP_FRAG_NOT_FOUND` returned here (meaning we override it with some other error code). I think we should fix that to ensure we always report the event via `send_drop_notify`.

jibi replied:
This should now be addressed by the second commit

pchaigno reviewed (changes requested):
I think we also need to extend the existing e2e fragment tracking test (`K8sServicesTest Checks service across nodes Supports IPv4 fragments`) to ensure we don't miss events or count them several times.

jibi commented on test/k8sT/Services.go:
```diff
@@ -1269,13 +1272,56 @@ var _ = Describe("K8sServicesTest", func() {
+
+			Expect(fragmentedPacketsAfter-fragmentedPacketsBefore).Should(Equal(fragments), "Failed to account for IPv4 fragments in BPF metrics")
+
+			// TODO: "Fragmented packet entry update failed" metric
```
I'm still looking into this, perhaps I can keep it for a subsequent PR?

I have some Go code to create IP packets:

```go
	ipLayer := layers.IPv4{
		Version: 4,
		Id:      4,
		//Flags:      layers.IPv4MoreFragments,
		//FragOffset: 64,
		Protocol: layers.IPProtocolUDP,
		SrcIP:    net.ParseIP("192.168.1.42"),
		DstIP:    net.ParseIP("192.168.1.1"),
	}
	udpLayer := layers.UDP{
		SrcPort: layers.UDPPort(1234),
		DstPort: layers.UDPPort(1234),
	}
	udpLayer.SetNetworkLayerForChecksum(&ipLayer)

	buf := gopacket.NewSerializeBuffer()
	opts := gopacket.SerializeOptions{
		FixLengths:       true,
		ComputeChecksums: true,
	}

	err := gopacket.SerializeLayers(buf, opts, &ipLayer, &udpLayer, gopacket.Payload("data"))
	if err != nil {
		panic(err)
	}

	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_RAW)
	if err != nil {
		panic(err)
	}
	defer syscall.Close(fd)

	err = syscall.Sendto(fd, buf.Bytes(), 0, &syscall.SockaddrInet4{})
	if err != nil {
		panic(err)
	}
```
but it looks like raw sockets don't play well with fragmentation as the moment I set the MF flag or the frag offset field, I stop seeing the packet on tcpdump :thinking:

Before resorting to iptables (which is something I'd like to avoid) I'll try to figure out if I can somehow make it work with raw sockets.

qmonnet replied:
No objection from my side :+1:

qmonnet replied:
Thinking again on the map update failure to test, we can maybe create a LRU map frozen on the BPF side prior to starting the Cilium agent, and make sure that it gets reused by the eBPF program. I think that could work. Could be part of the follow-up too.

pchaigno replied:
I think it's fine if we don't test that one metric. I doubt we tests the other equivalent metrics in the datapath and, in any case, we have to consider whether it's worth the additional test code complexity.

qmonnet commented on bpf/lib/metrics.h:
```diff
@@ -42,4 +42,23 @@ static __always_inline void update_metrics(__u32 bytes, __u8 direction,
+ * ct_to_metrics_dir
+ * @ct_dir: CT direction. Should be either CT_INGRESS or CT_EGRESS.
+ *
+ * Convert a CT_* direction into the corrisponding METRIC one.
```
```suggestion
 * Convert a CT_* direction into the corresponding METRIC one.
```

qmonnet commented on test/bpf/unit-test.c:
```diff
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: GPL-2.0
 // Copyright (c) 2018-2019 Authors of Cilium
 
+#define SKIP_UNDEF_LPM_LOOKUP_FN
```
Please add a comment to explain why we have this macro here.

jibi replied:
Done :heavy_check_mark:

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -678,6 +678,9 @@ var _ = Describe("K8sServicesTest", func() {
 			), "Failed to account for IPv4 fragments to %s (out)", dstIP)
+
+			return delta
+
```
Nit: blank line

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -1269,13 +1272,56 @@ var _ = Describe("K8sServicesTest", func() {
+			nodePort := data.Spec.Ports[1].NodePort
+			serverPort := data.Spec.Ports[1].TargetPort.IntValue()
+
+			cmd := "cilium bpf metrics list | grep 'Fragmented packet' | grep INGRESS | awk '{print $4}'"
```
Nit: Since you invoke awk already, you might as well avoid spawning grep twice? Something like (not tested):
```suggestion
			cmd := "cilium bpf metrics list | awk '/Fragmented packet.*INGRESS/ {print $4}'"
```
Ideally, we would even go for `-o jsonpath={...}`, but I had a quick look at the JSON output for that command and it looks... original. Let's stick plain output + awk for now, I guess.

jibi replied:
Nice, I should start learning awk for real :) 

> (not tested

(tested, works fine)

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -111,7 +112,12 @@ ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
 		return ret;
 
-	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
+	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY) < 0)
```
```suggestion
	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY))
```

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -111,7 +112,12 @@ ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
 
-	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
+	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY) < 0)
+		if (dir == CT_INGRESS || dir == CT_EGRESS)
```
This check is not needed? You already have it in `ct_to_metrics_dir()`.

jibi replied:
I think we still need this check as I'm guarding against the case where `dir` is equal to [`CT_SERVICE`](https://github.com/cilium/cilium/blob/master/bpf/lib/common.h#L536) (otherwise we may and up emitting a metric with the direction set to 0, which would be invalid)

qmonnet replied:
Ok fair enough

pchaigno replied:
Why are we ignoring all `CT_SERVICE` packets? They do correspond to fragmented ingressing and egressing packets, right? If they are already counted elsewhere, then do we even need to update the `IPV4_FRAG_DATAGRAMS_MAP` map?

qmonnet commented on bpf/lib/ipv4.h:
```diff
@@ -131,14 +137,18 @@ ipv4_handle_fragment(struct __ctx_buff *ctx,
 		.pad = 0,
 	};
 
+	if (dir == CT_INGRESS || dir == CT_EGRESS)
```
Same remark

qmonnet reviewed (changes requested):
Looks great, thanks!

A few nits inline, please also double-check your first commit log (I don't think `DROP_FRAG_NOT_FOUND` is added in that commit?)

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1956,6 +2001,7 @@ var _ = Describe("K8sServicesTest", func() {
 		SkipItIf(helpers.DoesNotRunOnNetNextOr419Kernel, "Supports IPv4 fragments", func() {
 			DeployCiliumAndDNS(kubectl, ciliumFilename)
 			testIPv4FragmentSupport()
+			testIPv4FragmentMetrics()
```
I wonder if we wouldn't benefit from merging `testIPv4FragmentSupport` and `testIPv4FragmentMetrics`. Right now, `testIPv4FragmentMetrics` checks fragmentation metrics for a single path, while `testIPv4FragmentSupport` already has tests for several different paths (via ClusterIP, with IPv4/IPv6, to remote and local nodes). If we merge the two functions, we could check metrics after each fragmented request and would avoid duplicated requests.

pchaigno reviewed (approved):
LGTM, but I'd still like to get some service-handling :eyes: (Daniel or Martynas) on this `CT_SERVICE` case.

> I would say to ignore the CT_SERVICE case for now for that metric (as this case never happens)

It doesn't happen _in tests_.

jibi committed:
bpf: add support for SERVICE metrics

This change introduces a new bpf metric (`METRIC_SERVICE`) which maps to
the CT_SERVICE direction of a flow.

Moreover, instead of having 2 different sets of constants for CT and
metrics directions (requiring some logic to convert one into another),
change the metrics directions so that they match the CT ones.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

christarazi reviewed (approved):
CI changes LGTM

brb commented on bpf/lib/conntrack.h:
```diff
@@ -527,14 +528,17 @@ static __always_inline int ipv4_ct_extract_l4_ports(struct __ctx_buff *ctx,
 
 	if (unlikely(ipv4_is_fragment(ip4)))
-		return ipv4_handle_fragment(ctx, ip4, off,
+		return ipv4_handle_fragment(ctx, ip4, off, dir,
```
Nit: maybe s/dir/ct_dir/ (also in other invocations too)?

brb reviewed (approved):
The `CT_SERVICE` case makes sense to me - in the case of a NodePort svc being accessed from outside and a ClusterIP svc from inside the cluster when bpf_sock is disabled, the LB lookup does a CT lookup with `dir=CT_SERVICE`.

Thanks, just one nit.

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -117,7 +118,10 @@ ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off,
 		return ret;
 
-	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
+	if (map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY))
```
We should mark this as `unlikely`.

pchaigno commented on bpf/lib/metrics.h:
```diff
@@ -56,6 +56,8 @@ static __always_inline int ct_to_metrics_dir(int ct_dir)
 	case CT_EGRESS:
 		return METRIC_EGRESS;
+	case CT_SERVICE:
+		return METRIC_SERVICE;
```
Could we reassign the `CT_XXX` values so they match our `METRICS_XXX` values 1:1 or is that already considered part of the API somehow? Doing so may help reduce complexity here. It's possible the compiler optimizes those statements enough that it doesn't matter; if the change isn't trivial, we might need to check the compiled instructions.

/cc @brb

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -527,14 +528,17 @@ static __always_inline int ipv4_ct_extract_l4_ports(struct __ctx_buff *ctx,
 	/* load sport + dport into tuple */
-	return ctx_load_bytes(ctx, off, &tuple->dport, 4);
+	if (ctx_load_bytes(ctx, off, &tuple->dport, 4) < 0)
+		return DROP_CT_INVALID_HDR;
```
Can `ctx_load_bytes` fail if we already checked the packet size earlier (`revalidate_data` in case `ENABLE_IPV4_FRAGMENTS` is defined)? If it cannot fail, maybe we could unconditionally return `CTX_ACT_OK` to reduce complexity?

Alternatively, we could also always perform the `revalidate_data` call and replace the `ctx_load_bytes` call by a direct packet read.

pchaigno reviewed (commented):
The smoke test caught a verifier complexity issue so I had a look at how we could reduce said complexity. The good news is that there aren't so many code changes that are susceptible of increasing complexity.

I suggested two potential changes below that may help reduce complexity if they can be implemented. The addition of `unlikely` may also affect complexity but it seems less likely.

jibi committed:
bpf: optimize relax_verifier()

Replace the call to `csum_diff` with `get_smp_processor_id()` as with
the latter we can avoid having to init r1-r5 registers.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
bpf: reintroduce relax_verifier()

relax_verifier() is a dummy helper call to introduce a pruning
checkpoint to help relax the verifier to avoid reaching complexity
limits on older kernels.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

joestringer commented on bpf/lib/common.h:
```diff
@@ -395,14 +402,17 @@ enum {
+#define METRIC_EGRESS   0
 #define METRIC_INGRESS  1
-#define METRIC_EGRESS   2
+#define METRIC_SERVICE  2
```
I don't understand, what direction is service? I don't see any other use of this value in the PR either.

jibi replied:
Here I'm updating the `METRIC_*` direction values to have a consistent 1:1 mapping with the CT ones, so `METRIC_SERVICE` is used for packets with a `CT_SERVICE` direction.

As Martynas noted:
> in the case of a NodePort svc being accessed from outside and a ClusterIP svc from inside the cluster when bpf_sock is disabled, the LB lookup does a CT lookup with dir=CT_SERVICE.

joestringer replied:
For CT, we create independent entries depending on whether it's a direction or if it's a service. For metrics, we are only trying to observe traffic in a particular direction. I don't think it's helpful to expose CT internal entry type details onto the metrics directionality which is intended for user consumption and interpretation of tx/Rx packets.

I think I missed the point where we were trying to take advantage of having the same numeric values for CT and Metric constants, would you mind pointing to it? If we find it useful to converge these constant values for some reason, I'm not against that. At the same time, I think it will make metrics direction more confusing if we assume we can just forward the CT entry type as a direction up through to the user.

jibi replied:
(sorry, I somehow missed this comment yesterday)

> I think I missed the point where we were trying to take advantage of having the same numeric values for CT and Metric constants, would you mind pointing to it?

It started from here: https://github.com/cilium/cilium/pull/13347#issuecomment-707333299

There are 2 things actually:
* first `ipv4_handle_fragmentation()` needs to know the direction of the packet in order to emit metrics. The source for this is the `dir` parameter of `ct_lookup4()`. The initial problem was that this `dir` variable would be sometimes set to `CT_SERVICE`, so it was not obvious how to map it into `METRIC_INGRESS`/`METRIC_EGRESS` (so I proposed to introduce a new `METRIC_SERVICE`).
* next, since we now have 2 sets of constants, `METRIC_{INGRESS, EGRESS, SERVICE}` and `CT_{INGRESS, EGRESS, SERVICE}`, with inconsistent values (e.g. `METRIC_EGRESS` did not match `CT_EGRESS`) I proposed to converge these constant values (in order to get rid of a helper to convert one direction into another, which should reduce the complexity of the program).

joestringer replied:
I see, and I guess it's difficult to pass the actual direction down from higher up the stack due to the number of layers involved and some of the functions also having many parameters.. Hmm.

No solutions come to mind, not sure there's much we can do. 

:+1:

joestringer commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -48,19 +48,23 @@ const (
-	dirUnknown = 0
+	dirService = 2
+
+	dirUnknown = 255
```
FWIW the directionality here used 0 as the "unknown" value in case there were any paths that logged metrics if the value was left uninitialized / as zero. I'm not sure if there's actually any paths left that would use that value for the direction.

If unknown isn't 0, I think there's no point in the unknown value. I don't think any path in the datapath will accidentally set the value to 255.

jibi committed:
bpf: fix complexity issue on kernels <5.3

This commit reduces the complexity of the "to-container" section by
introducing a few state pruning points with the help of relax_verifier().

Pruning points have been determined by looking at the instructions that
the verifier is spending the most passes on.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
bpf: add metrics for fragmented IPv4 packets

This commit introduces 2 new metrics in the datapath logic related to
fragmented IPv4 packets:

* `REASON_FRAG_PACKET`: number of received fragmented packets
* `REASON_FRAG_PACKET_UPDATE`: number of failures in updating the
  `IPV4_FRAG_DATAGRAMS_MAP` map to register the first logical fragment of
  a datagram

Fixes: #11179

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
bpf: reduce complexity of logic to handle IPv4 fragments

Refactor the logic to handle IPv4 fragments to have only a single call
to `ctx_load_bytes()` in order to load the l4 ports tuple.

This change fixes a complexity issue in 5.4 kernels which prevents the
BPF program from being loaded by the verifier as it would reach the
maximum amount of instructions that can be processed.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi commented on bpf/include/bpf/verifier.h:
```diff
@@ -0,0 +1,20 @@
+static __always_inline void relax_verifier(void)
+{
+#ifdef NEEDS_RELAX_VERIFIER
+       volatile int __maybe_unused id = get_smp_processor_id();
```
Following up on last week discussion on the performance implications of using `ktime_get_ns()` I have switched to `get_smp_processor_id()` which should get compiled down to a single instruction (at least on x86-64).

To confirm that I had look at the generated asm.

test module:
```c
..
int init_module(void)
{
	volatile int x = smp_processor_id();
	return 0;
}
```

objdump output:
```
0000000000000000 <init_module>:
   0:	e8 00 00 00 00       	callq  5 <init_module+0x5>
			1: R_X86_64_PLT32	__fentry__-0x4
   5:	55                   	push   %rbp
   6:	48 89 e5             	mov    %rsp,%rbp
   9:	48 83 ec 08          	sub    $0x8,%rsp
   d:	65 8b 05 00 00 00 00 	mov    %gs:0x0(%rip),%eax        # 14 <init_module+0x14>
			10: R_X86_64_PC32	cpu_number-0x4
  14:	89 45 fc             	mov    %eax,-0x4(%rbp)
..
```

pchaigno replied:
/cc @borkmann

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -102,6 +104,8 @@ var (
 
 	bootstrapTimestamp = time.Now()
 
+	needsRelaxVerifierKernelVer = versioncheck.MustCompile("<5.3.0")
```
We usually probe for kernel features using `probe.NewProbeManager()` (see `bpf_sk_assign` in the same file). That relies on `bpftool feature probe macro`. If the feature we need isn't exposed by bpftool, then we usually rely on some helper introduced around the same time and add the wanted feature probe to https://github.com/cilium/cilium/issues/11014.

In your case, maybe `HAVE_LARGE_INSN_LIMIT` is a good fit? IIRC, that was introduced around the same time, maybe even in the same batch?

jibi replied:
Yup, I exposed the `have_large_insn_limit` feature which tests the presence of the `c04c0d2b968a ("bpf: increase complexity limit and maximum program size")` commit :+1:

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -143,20 +138,32 @@ ipv4_handle_fragment(struct __ctx_buff *ctx,
+			return ipv4_frag_get_l4ports(&frag_id, ports);
+	}
+
+	ret = ctx_load_bytes(ctx, l4_off, ports, 4);
```
Can we keep the `/* load sport + dport into tuple */` comment? It's not super obvious that we are loading both on a quick read.

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -108,16 +108,10 @@ ipv4_frag_get_l4ports(const struct ipv4_frag_id *frag_id,
 
 static __always_inline int
-ipv4_frag_register_datagram(struct __ctx_buff *ctx, int l4_off, int dir,
+ipv4_frag_register_datagram(struct __ctx_buff *ctx, int dir,
```
Probably worth getting rid of that function now.

pchaigno reviewed (approved):
Only minor comments from me. Nice work @jibi! :tada: :ship:

kaworu commented on bpf/lib/common.h:
```diff
@@ -328,6 +328,9 @@ enum {
  * where reason > 0 corresponds to -(DROP_*)
+ *
+ * These are shared with pkg/monitor/api/drop.go, when modifying any of the
+ * below, that file should also be updated.
```
On the way, could you add a comment in _pkg/monitor/api/drop.go_ `errors` that DROP values are also shared in _api/v1/flow/flow.proto_ `enum DropReason` please?

kaworu commented on test/k8sT/Services.go:
```diff
@@ -818,6 +818,14 @@ var _ = Describe("K8sServicesTest", func() {
 				countOutK8s2, _ = strconv.Atoi(strings.TrimSpace(res.Stdout()))
 			}
 
+			cmdFragmentedPacketsBeforeK8s1 := "cilium bpf metrics list | awk '/Fragmented packet.*INGRESS/ {print $4}'"
```
I think this command deserve a func in _test/helpers/wrappers.go_ to avoid its duplication below.

kaworu commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -124,12 +126,10 @@ func (k *Key) String() string {
 		return direction[dir]
 	}
-	return direction[dirUnknown]
+	return "UNKNOWN"
```
what about simply
```go
if desc, ok := direction[dir]; ok {
    return desc
}
return "UNKNOWN"
```

jibi committed:
datapath/probes: add support for misc features

This commit adds support for the "misc" features reported by bpftool.
This is exposed by the ProbeManager object through the GetMisc() method.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

kaworu commented on bpf/lib/common.h:
```diff
@@ -387,6 +390,9 @@ enum {
  * If reason is larger than below then this is a drop reason and
  * value corresponds to -(DROP_*), see above.
+ *
+ * These are shared with pkg/monitor/api/drop.go and api/v1/flow/flow.proto.
```
nit: _api/v1/flow/flow.proto_ doesn't expose the "non-drop reasons" (i.e. <130), so it can be removed from this comment.

```suggestion
 * These are shared with pkg/monitor/api/drop.go.
```

kaworu reviewed (approved):
Great work @jibi! I only have one more minor comment suggestion :rocket:

christarazi commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -131,11 +131,17 @@ type MapTypes struct {
 	HaveStackMapType               bool `json:"have_stack_map_type"`
 }
 
+// Misc contains bools exposing miscellaneous eBPF features.
```
Might be helpful to say where these are coming from, i.e. "bpftool ..."

jibi replied:
I think it would be redundant as that's already in the comment of the `Features` struct (which contains `Misc`)

https://github.com/cilium/cilium/blob/10d2ea7699492a14ca98d4424eb5a946077efdb5/pkg/datapath/linux/probes/probes.go#L139-L145
